src/grading.asm

```diff
@@ -1059,38 +1108,539 @@ UpdateGradeSHIR:
+    ; Make HL point to the table that contains the level multipliers
+    ld hl, sTGM3LevelMultiplier
+    ; Get our level into BC
+    ld a, [hCLevel+3]
+    ld b, a
+    ld a, [hCLevel+2]
+    swap a
+    or b
+    ld c, a
+    ld a, [hCLevel+1]
+    ld b, a
+    ld a, [hCLevel]
+    swap a
+    or b
+    ld b, a
+.Level750
+    ; Is our level 750 or higher?
+    ld a, b
+    cp a, LEVEL_MULT_3A
+    ; If B is less than 7, that means we are not even in level 700, so check for 500
+    jr c, .Level500
+    ; If B is NOT less than 7, we might be in level 750 or greater, so check the remaining digits.
+    ld a, c
+    cp a, LEVEL_MULT_3B
+    ; If C is less than 50, we didn't reach level 750 yet, but we are for sure in the 7** Section, load the corresponding offset.
+    jp .under750
+    ; If C is equal or greater than 50, then congrats!, load the corresponding offset
+    ld b, 0
+    ld c, 2 
+    add hl, bc
+    ld a, [hl]
+    jp .Multiply
+.under750
+    ld b, 0
+    ld c, 1
+    add hl, bc
+    ld a, [hl]
+.Level500
+    ; Is our level 500 or higher?
+    ld a, b
+    cp a, LEVEL_MULT_2A
+    ; If B is less than 5, that means we are not even in level 500, so check for 250
+    jr c, .Level250
+    ; If B is NOT less than 5, we are in level 500 or greater
+    ld b, 0
+    ld c, 2 
+    add hl, bc
+    ld a, [hl]
+    jp .Multiply
+.Level250 ; There is no Offset, so just get the multiplier
+    ; Is our level 750 or higher?
+    ld a, b
+    cp a, LEVEL_MULT_1A
+    ; If B is less than 2, that means we are not even in level 200, so no multiplier
+    jr c, .under250
+    ; If B is NOT less than 2, we might be in level 250 or greater, so check the remaining digits.
+    ld a, c
+    cp a, LEVEL_MULT_1B
+    ; If C is less than 50, we didn't reach level 250 yet, so no multiplier
+    jp .under250
+    ; If C is equal or greater than 50, then congrats!, load the corresponding offset (I said there is no Offset!)
+    ld a, [hl]
+    jp .Multiply
+.under250 ; There is no multiplier, so just load the points
+    ld a, e
+    jp .loadpoints
+.Multiply ; FINALLY!!!!!, This took forever!
+    ld b, a ; Load the multiplier into B
+    ld a, e ; Remember the points we got earlier?, here they are... Again.
+    ld c, a ; We will add from C
+:   add a, c
+    dec b
+    jr nz, :-
+    ; AND NOW WE ARE DONE!, LOAD THOSE POINTS!
+    jp .loadpoints
+.IncreaseInternalGrade
+    ; Are we on level *00-*05?
+    ld a, [hCLevel+CLEVEL_ONES]
+    cp a, 6
+    ; If we are, jump to the update COOL grade funcion just in case we have to apply a section COOL
+    jr c, CheckCOOL
+    ; If not, continue
+    ; Do we have 100 Grade Points?
+    ld a, [wInternalGradePoints]
+    cp a, 100
+    ret c ; If the Internal Grade Points is less than 100, return, we don't have to change our Grade
+    xor a, a ; Reset the points to 0 and increase the internal grade
+    ld [wInternalGradePoints], a
+    ld a, [wInternalGrade]
+    inc a
+    ld [wInternalGrade], a

```

Comment:
It is a good idea to leave a comment when one function falls through to the next, so that I know it's not a mistake.